lib/src/voip/group_call_session.dart

```diff
@@ -0,0 +1,1361 @@
+import 'package:matrix/src/voip/models/call_membership.dart';
+import 'package:matrix/src/voip/models/call_options.dart';
+import 'package:matrix/src/voip/models/voip_id.dart';
+import 'package:matrix/src/voip/utils/stream_helper.dart';
+
+/// Holds methods for managing a group call. This class is also responsible for
+/// holding and managing the individual `CallSession`s in a group call.
+class GroupCallSession {
+  // Config
+  static const updateExpireTsTimerDuration = Duration(seconds: 15);
+  static const expireTsBumpDuration = Duration(seconds: 45);
+  static const activeSpeakerInterval = Duration(seconds: 5);
+
+  final Client client;
+  final VoIP voip;
+  final Room room;
+
+  /// is a list of backend to allow passing multiple backends in the future
+  /// we use the first backend everywhere as of now
+  final List<CallBackend> backends;
+  final String? application;
+  final String? scope;
+
+  GroupCallState state = GroupCallState.localCallFeedUninitialized;
+  StreamSubscription<CallSession>? _callSubscription;
+
+  /// participant:volume
+  final Map<CallParticipant, double> audioLevelsMap = {};
+  CallParticipant? activeSpeaker;
+  WrappedMediaStream? localUserMediaStream;
+  WrappedMediaStream? localScreenshareStream;
+  String? localDesktopCapturerSourceId;
+  List<CallSession> callSessions = [];
+
+  CallParticipant? get localParticipant => voip.localParticipant;
+
+  /// userId:deviceId
+  List<CallParticipant> participants = [];
+  List<WrappedMediaStream> userMediaStreams = [];
+  List<WrappedMediaStream> screenshareStreams = [];
+  late String groupCallId;
+
+  GroupCallError? lastError;
+
+  Timer? activeSpeakerLoopTimeout;
+
+  Timer? resendMemberStateEventTimer;
+  Timer? memberLeaveEncKeyRotateDebounceTimer;
+
+  final CachedStreamController<GroupCallSession> onGroupCallFeedsChanged =
+      CachedStreamController();
+
+  final CachedStreamController<GroupCallState> onGroupCallState =
+      CachedStreamController();
+
+  final CachedStreamController<GroupCallEvent> onGroupCallEvent =
+      CachedStreamController();
+
+  final CachedStreamController<WrappedMediaStream> onStreamAdd =
+      CachedStreamController();
+
+  final CachedStreamController<WrappedMediaStream> onStreamRemoved =
+      CachedStreamController();
+
+  bool get isLivekitCall => backends.first is LiveKitBackend;
+
+  /// toggle e2ee setup and key sharing
+  final bool enableE2EE;
+
+  GroupCallSession({
+    String? groupCallId,
+    required this.client,
+    required this.room,
+    required this.voip,
+    required this.backends,
+    required this.enableE2EE,
+    this.application = 'm.call',
+    this.scope = 'm.room',
+  }) {
+    this.groupCallId = groupCallId ?? genCallID();
+  }
+
+  String get avatarName =>
+      getUser().calcDisplayname(mxidLocalPartFallback: false);
+
+  String? get displayName => getUser().displayName;
+
+  User getUser() {
+    return room.unsafeGetUserFromMemoryOrFallback(client.userID!);
+  }
```

Comment:
Should we really expose an unsafe function as safe? Is it safe in this context?

Reply:
fixed